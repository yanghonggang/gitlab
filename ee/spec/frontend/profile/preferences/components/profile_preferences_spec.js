import { shallowMount } from '@vue/test-utils';

import GroupOverviewSelector from 'ee/profile/preferences/components/group_overview_selector.vue';
import ProfilePreferences from '~/profile/preferences/components/profile_preferences.vue';
import { groupViewChoices } from '../mock_data';

describe('ProfilePreferences component', () => {
  let wrapper;
  const defaultProvide = {
    firstDayOfWeekChoicesWithDefault: [],
    dashboardChoices: [],
    layoutChoices: [],
    languageChoices: [],
    projectViewChoices: [],
    groupViewChoices: [],
    integrationViews: [],
    userFields: {},
    featureFlags: {},
  };

  function createComponent(options = {}) {
    const { props = {}, provide = {} } = options;
    return shallowMount(ProfilePreferences, {
      provide: {
        ...defaultProvide,
        ...provide,
      },
      propsData: props,
      stubs: {
        GroupOverviewSelector,
      },
    });
  }

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  describe('Group view section', () => {
    it('should render the component', () => {
      wrapper = createComponent();
      const groupOverviewSelector = wrapper.find(GroupOverviewSelector);
      expect(groupOverviewSelector.exists()).toBe(true);
    });

    describe('without data', () => {
      beforeEach(() => {
        wrapper = createComponent();
      });

      it('should not render the heading', () => {
        const groupOverviewHeading = wrapper.find(
          '[data-testid="profile-preferences-group-overview-heading"]',
        );
        expect(groupOverviewHeading.exists()).toBe(false);
      });
      it('should not render the options', () => {
        const groupOverviewOptions = wrapper.findAll(
          '[data-testid="profile-preferences-group-overview-option"]',
        );
        expect(groupOverviewOptions).toHaveLength(0);
      });
    });

    describe('with data', () => {
      beforeEach(() => {
        wrapper = createComponent({ provide: { groupViewChoices } });
      });

      it('should render the heading', () => {
        const groupOverviewHeading = wrapper.find(
          '[data-testid="profile-preferences-group-overview-heading"]',
        );
        expect(groupOverviewHeading.exists()).toBe(true);
      });

      it('should render the options', () => {
        const groupOverviewOptions = wrapper.findAll(
          '[data-testid="profile-preferences-group-overview-option"]',
        );
        expect(groupOverviewOptions).toHaveLength(groupViewChoices.length);
      });
    });
  });

  it('should render ProfilePreferences properly', () => {
    wrapper = createComponent({
      provide: { groupViewChoices },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
