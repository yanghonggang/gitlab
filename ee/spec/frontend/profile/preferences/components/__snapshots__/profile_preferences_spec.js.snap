// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePreferences component should render ProfilePreferences properly 1`] = `
<div
  class="row gl-mt-3 js-preferences-form"
>
  <div
    class="col-sm-12"
  >
    <hr />
  </div>
   
  <div
    class="col-lg-4 profile-settings-sidebar"
  >
    <h4
      class="gl-mt-0"
    >
      
      Behavior
    
    </h4>
     
    <p>
      
      This setting allows you to customize the behavior of the system layout and default views.
      
      <a
        href="/help/user/profile/preferences#behavior"
        target="_blank"
      >
        
        Learn more 
      </a>
      .
    
    </p>
  </div>
   
  <div
    class="col-lg-8"
  >
    <div
      class="form-group"
    >
      <label
        class="label-bold"
        for="user_layout"
      >
        
        Layout width
      
      </label>
       
      <div
        class="select-wrapper"
      >
        <select
          class="form-control select-control"
          id="user_layout"
          name="user[layout]"
          tabindex="-1"
          title="Layout width"
        />
         
        <i
          aria-hidden="true"
          class="fa fa-chevron-down"
        />
      </div>
       
      <div
        class="form-text text-muted"
      >
        
        Choose between fixed (max. 1280px) and fluid (100%) application layout.
      
      </div>
    </div>
     
    <div
      class="form-group"
    >
      <label
        class="label-bold"
        for="user_dashboard"
      >
        
        Homepage content
      
      </label>
       
      <div
        class="select-wrapper"
      >
        <select
          class="form-control select-control"
          id="user_dashboard"
          name="user[dashboard]"
          tabindex="-1"
          title="Homepage content"
        />
         
        <i
          aria-hidden="true"
          class="fa fa-chevron-down"
        />
      </div>
       
      <div
        class="form-text text-muted"
      >
        
        Choose what content you want to see on your homepage.
      
      </div>
    </div>
     
    <div
      class="form-group"
    >
      <label
        class="label-bold"
        data-testid="profile-preferences-group-overview-heading"
        for="user_group_view"
      >
        
    Group overview content
  
      </label>
       
      <div
        class="select-wrapper"
      >
        <select
          class="form-control select-control"
          name="user[group_view]"
          tabindex="-1"
          title="Group overview content"
        >
          <option
            data-testid="profile-preferences-group-overview-option"
            value="details"
          >
            
        Details
      
          </option>
          <option
            data-testid="profile-preferences-group-overview-option"
            value="security"
          >
            
        Security Dashboard
      
          </option>
        </select>
         
        <i
          aria-hidden="true"
          class="fa fa-chevron-down"
        />
      </div>
       
      <div
        class="form-text text-muted"
      >
        
    Choose what content you want to see on a group’s overview page.
  
      </div>
    </div>
     
    <div
      class="form-group"
    >
      <label
        class="label-bold"
        for="user_project_view"
      >
        
        Project overview content
      
      </label>
       
      <div
        class="select-wrapper"
      >
        <select
          class="form-control select-control"
          id="user_project_view"
          name="user[project_view]"
          tabindex="-1"
          title="Project overview content"
        />
         
        <i
          aria-hidden="true"
          class="fa fa-chevron-down"
        />
      </div>
       
      <div
        class="form-text text-muted"
      >
        
        Choose what content you want to see on a project’s overview page.
      
      </div>
    </div>
     
    <div
      class="form-group form-check"
    >
      <input
        name="user[render_whitespace_in_code]"
        type="hidden"
        value="0"
      />
       
      <input
        class="form-check-input"
        id="user_render_whitespace_in_code"
        name="user[render_whitespace_in_code]"
        type="checkbox"
        value="1"
      />
       
      <label
        class="form-check-label"
        for="user_render_whitespace_in_code"
      >
        
        Render whitespace characters in the Web IDE
      
      </label>
    </div>
     
    <div
      class="form-group form-check"
    >
      <input
        name="user[show_whitespace_in_diffs]"
        type="hidden"
        value="0"
      />
       
      <input
        class="form-check-input"
        id="user_show_whitespace_in_diffs"
        name="user[show_whitespace_in_diffs]"
        type="checkbox"
        value="1"
      />
       
      <label
        class="form-check-label"
        for="user_show_whitespace_in_diffs"
      >
        
        Show whitespace changes in diffs
      
      </label>
    </div>
     
    <!---->
     
    <div
      class="form-group"
    >
      <label
        class="label-bold"
        for="user_tab_width"
      >
        
        Tab width
      
      </label>
       
      <input
        class="form-control"
        id="user_tab_width"
        max="12"
        min="1"
        name="user[tab_width]"
        required="required"
        type="number"
      />
       
      <div
        class="form-text text-muted"
      >
        
        Must be a number between 1 and 12
      
      </div>
    </div>
  </div>
   
  <div
    class="col-sm-12"
  >
    <hr />
  </div>
   
  <div
    class="col-lg-4 profile-settings-sidebar"
  >
    <h4
      class="gl-mt-0"
    >
      
      Localization
    
    </h4>
     
    <p>
      
      Customize language and region related settings.
      
      <a
        href="/help/user/profile/preferences#localization"
        target="_blank"
      >
        
        Learn more 
      </a>
      .
    
    </p>
  </div>
   
  <div
    class="col-lg-8"
  >
    <div
      class="form-group"
    >
      <label
        class="label-bold"
        for="user_preferred_language"
      >
        
        Language
      
      </label>
       
      <div
        class="select-wrapper"
      >
        <select
          class="form-control select-control"
          id="user_preferred_language"
          name="user[preferred_language]"
          tabindex="-1"
          title="Language"
        />
         
        <i
          aria-hidden="true"
          class="fa fa-chevron-down"
        />
      </div>
       
      <div
        class="form-text text-muted"
      >
        
        This feature is experimental and translations are not complete yet
      
      </div>
    </div>
     
    <div
      class="form-group"
    >
      <label
        class="label-bold"
        for="user_first_day_of_week"
      >
        
        First day of the week
      
      </label>
       
      <div
        class="select-wrapper"
      >
        <select
          class="form-control select-control"
          id="user_first_day_of_week"
          name="user[first_day_of_week]"
          tabindex="-1"
          title="First day of the week"
        />
         
        <i
          aria-hidden="true"
          class="fa fa-chevron-down"
        />
      </div>
    </div>
  </div>
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
</div>
`;
